release/release.go

```diff
@@ -142,3 +148,39 @@ func PreRelease(repo *gogit.Repository, options PreReleaseOptions) PreReleaseFun
 		return pre, nil
 	}
 }
+
+func isCommitOnBranch(repo *gogit.Repository, commit plumbing.Hash, branch plumbing.ReferenceName) (bool, error) {
+	branchRef, err := repo.Reference(branch, true)
+	if err != nil {
+		return false, err
+	}
+
+	reaches, err := reaches(repo, branchRef.Hash(), commit)
+	if err != nil {
+		return false, err
+	}
+
+	return reaches, nil
+}
+
+func reaches(repo *gogit.Repository, branchCommit, commitToFind plumbing.Hash) (bool, error) {
+	if branchCommit == commitToFind {
+		return true, nil
+	}
+
+	branchCommitObject, err := repo.CommitObject(branchCommit)
+	if err != nil {
+		return false, err
+	}
+
+	for _, parentHash := range branchCommitObject.ParentHashes {
+		reaches, err := reaches(repo, parentHash, commitToFind)
```

Comment:
this recursive call will blow the stack.  i think you need a loop of some sort.

Reply:
I'm not sure how you'd get this to work with just a loop. `go-git` doesn't have a feature to loop through commits in a branch. I'm also not sure it'd be possible to convert this to tail recursion, but I've only briefly looked.

FWIW I ran this against a branch off of `core-platform` which has ~16,000 commits with no issues and it returned almost immediately.

Reply:
Actually I'm not sure my manual test against core-platform was correct. Re-running you can see it taking a long time.

Reply:
> FWIW I ran this against a branch off of core-platform which has ~16,000 commits with no issues and it returned almost immediately.

Did you run it with a non-reachable commit ID?  That's when the stack will blow (because it has to check all 16,000 commits on the branch before it can be satisfied that the commit is not reachable, and each commit pushes a new stack frame)

> I'm also not sure it'd be possible to convert this to tail recursion

No(t easily) -- also the go language doesn't mandate support for optimized tail recursion so even that wouldn't help.

> go-git doesn't have a feature to loop through commits in a branch

No, you have to do it yourself.  The "dumb way" is with a deque of commits where you do:

```
while commit := deque.Pop() {
    if commit == commitToSearchFor {
        return found
    }
   deque.Push(commit.Parents())
}
```

You can optimize this by staying out of the data structure when you are looking at linear history:

```
commitsToCheck := deque.New(initialCommit)
while commit := commitsToCheck.Pop() {
    while len(commit.Parents()) == 1 {
        if commit == commitToFind { return found }
        commit = commit.Parents()[0]
    }
    if commit == commitToFind { return found }
    commitsToCheck.Push(commit.Parents()...)
}
```

(also you probably want to figure out how to do it without a deque but using only slices, which probably can be used as a shitty deque anyway since allocation isn't a major concern as long as you apply the optimization for linear history)

Reply:
Whats frustrating, is that this is built into `git` cli. We could use that and it'd be as simple as:
```golang
func isCommitOnBranch(repo *gogit.Repository, commit plumbing.Hash, branch string) (bool, error) {
	cmd := exec.Command("bash", "-c", "git branch -r --contains ", commit.String())

	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()

	if err != nil {
		return false, errors.New("unable to check if git commit exists for branch:\n " + stderr.String())
	}

	branchesForCommit := strings.Split(stdout.String(), "\n")
	return contains(branchesForCommit, "origin/"+branch), nil
}
```

But I'm hesitant to go down that route given we already use `go-git`.

Reply:
If I introduce the memoization that was on the linked example, it runs much faster (~2s). 

```bash
16:36:25 [~/workspace/core-platform] {130}:git st
HEAD detached at c79eec274
nothing to commit, working tree clean
16:36:26 [~/workspace/core-platform] {0}:../vergo/bin/vergo-memorized check release --log-level=debug
DEBU[0000] Current branch:HEAD, short: HEAD
WARN[0001] Commit not found on branch [branch: refs/remotes/origin/master, head: c79eec2743671696d836e4d4de2e0e0709a96056, ref: 680ea2fc80029e4dcea251b0a91b27207d9599aa]
DEBU[0001] branchRef could not be resolved: refs/remotes/origin/main  error="reference not found"
Error: HEAD validation: invalid headless checkout
16:36:31 [~/workspace/core-platform] {1}:
```

With `c79eec274` being a revision of a commit on a branch of the `core-platform` repo. 

I'll push that up now.

It's also worth mentioning that this shouldn't happen often. Only if a team has misconfigured a cd script. Usually tenants will check they're on master/main before calling vergo.

Reply:
@aecay Does that alleviate your concerns? If not we'll have to get this bug prioritised within our sprints to dedicate more time to it. At the moment I've been squeezing this in and around meetings etc.

Reply:
yes, the new change lgtm.  thanks! 🙂